server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/VariableWidthHistogramAggregator.java

```diff
@@ -0,0 +1,338 @@
+            // The bucket doesn't exist yet -> Allocate space for it
+            clusterMaxes = bigArrays.resize(clusterMaxes, bucketOrd + 1);
+            clusterMins = bigArrays.resize(clusterMins, bucketOrd + 1);
+            clusterCentroids = bigArrays.resize(clusterCentroids, bucketOrd + 1);
+            clusterSizes = bigArrays.resize(clusterSizes, bucketOrd + 1);
+
+            bucketMax = val;
+            bucketMin = val;
+            clusterSizes.set(bucketOrd, 0);
+        } else {
+            bucketMax = Math.max(clusterMaxes.get(bucketOrd), val);
+            bucketMin = Math.min(clusterMins.get(bucketOrd), val);
+        }
+
+        // Recalculate the centroid
+        double oldCentroid = clusterCentroids.get(bucketOrd);
+        double clusterSize = clusterSizes.get(bucketOrd);
+        double newCentroid = ((oldCentroid * clusterSize) + val) / (clusterSize + 1);
+
+        clusterMaxes.set(bucketOrd, bucketMax);
+        clusterMins.set(bucketOrd, bucketMin);
+        clusterCentroids.set(bucketOrd, newCentroid);
+        clusterSizes.increment(bucketOrd, 1);
+    }
+
+    private void bucketCachedDocs() throws IOException{
+        Integer[] indexes = new Integer[(int)cachedValues.size()];
+        for(int i = 0; i < indexes.length; i++){
+            indexes[i] = i;
+        }
+
+        long[] mergeMap = new long[(int)cachedValues.size()];
+
+        // Sorting the documents by key enables bucketing with a single linear scan
+        // While sorting we have to track each document's initial index to determine which bucket its doc is in
+        // We can do this tracking by sorting the documents using their indexes
+        // After this step, this.cachedValues[this.indexes[i]] will be the i'th smallest value
+        Arrays.sort(indexes, Comparator.comparingDouble(idx -> cachedValues.get((long) idx)));
+
+        // An aggregator cannot return more than shardSize buckets.
+        // So leave room for some distant buckets to be added during rest of the collection phase.
+        int initialNumBuckets = shardSize * 3/4;
+
+        // Naively use basic linear separation to group the first cacheLimit docs into initialNumBuckets buckets
+        int docsPerBucket = (int) Math.ceil((double) cachedValues.size() / (double) initialNumBuckets);
+        int bucketOrd = 0;
+        for(int i = 0; i < indexes.length; i++){
+            // The doc that is currently in bucket indexes[i] is the i'th smallest
+            // We will move it into the the bucket at `bucketOrd`
+
+            int index = indexes[i];
+            double val = cachedValues.get(index);
+
+            mergeMap[index] = bucketOrd;
+            addToCluster(bucketOrd, val);
+
+            if((i + 1) % docsPerBucket == 0){
+                // There are enough docs in this bucket, so we should make a new one
+                bucketOrd += 1;
+            }
+        }
+
+        mergeBuckets(mergeMap, initialNumBuckets);
```

Comment:
I'm not sure I understand how swapping buckets is different from merging them. 

From reading through the method, It looks like `deferringCollector.mergeBuckets(mergeMap)` just applies the merge map. And since I'm able to encode the buckets swaps into a merge map, I don't see why it wouldn't work?

I've added a new test which requires a complex merge map to be applied on an aggregator with sub aggregations: `VariableWidthHistogramAggregatorTests.testComplexSubAggregations`. 

This test fails when run on the old version. However, now that I've implemented your suggested change it is passing.